Fix #1386: Add rewards reporting to support media publishers and activity

kylehickinson opened the pull request:
- Panel now shows proper channels, names, etc. for supported publishers (YouTube, Vimeo, Twitch, GitHub, Twitter)
- Panel now shows proper activity for video related publishers (YouTube, Vimeo, Twitch)
- Now "hides" the tab (therefore computing activity) when the rewards panel is shown and "shows" it when its hidden
- Does the same when full-screen modals appear (tab is technically "hidden")
- Shows correct verified badge for publishers
- Supports notification count, however UI isn't added yet to show the red dot yet

<!-- 
*Thank you for submitting a pull request, your contributions are greatly appreciated!*

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. 
List any dependencies that are required for this change.
-->

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [x] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)
- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New files have MPL-2.0 license header.


## Test Plan:

Prefix: Must remove `NO_REWARDS` from config for rewards to appear in the toolbar

1. Visit a normal publisher website, wait 8+ seconds
1. Open rewards panel: should show activity %
1. Visit a verified publisher website, should see badge in the toolbar
1. Open rewards panel: should show verified
1. Visit each video media publisher: YouTube, Vimeo, Twitch, load a video, and watch for 8+ seconds
1. Open rewards panel: should show `<user> on <platform>` as the name, activity should be non-zero
1. Visit a user profile of the other media publishers: GitHub, Twitter, Reddit and wait 8+ seconds on the page
1. Open rewards panel: should show `<user> on <platform>` as the name, **activity should be zero** as these publishers do not accumulate duration activity

### Screenshots:

| Twitter | Twitch |
| -- | -- |
| ![Simulator Screen Shot - iPhone 8 - 2019-08-13 at 12 25 11](https://user-images.githubusercontent.com/529104/62959148-1a10e580-bdc6-11e9-8f41-38ead36e42f3.png) | ![Simulator Screen Shot - iPhone 8 - 2019-08-13 at 12 29 29](https://user-images.githubusercontent.com/529104/62959164-2432e400-bdc6-11e9-86b8-d2e2b10b69f1.png) |
| YouTube | GitHub | 
| ![Simulator Screen Shot - iPhone 8 - 2019-08-13 at 12 28 21](https://user-images.githubusercontent.com/529104/62959176-2a28c500-bdc6-11e9-8e61-bd9c79bbf431.png) | ![Simulator Screen Shot - iPhone 8 - 2019-08-13 at 12 29 14](https://user-images.githubusercontent.com/529104/62959189-33199680-bdc6-11e9-99ef-2c317e949b3f.png) 
| Auto-Contribute | |
| ![Simulator Screen Shot - iPhone 8 - 2019-08-13 at 12 28 54](https://user-images.githubusercontent.com/529104/62959243-52182880-bdc6-11e9-9997-cdb2623d67ec.png) |

## Reviewer Checklist:

- [ ] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [ ] Issues are assigned to at least one epic.
- [ ] Issues include necessary QA labels:
  - [ ] `QA/(Yes|No)`
  - [ ] `release-notes/(include|exclude)`
  - [ ] `bug` / `enhancement`
- [ ] Necessary security reviews have taken place.
- [ ] Adequate test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable)

Brandon-T commented on Client/Assets/AllFramesAtDocumentEnd.js:
```diff
@@ -1 +1 @@
-!function(e){var n={};function t(r){if(n[r])return n[r].exports;var i=n[r]={i:r,l:!1,exports:{}};return e[r].call(i.exports,i,i.exports,t),i.l=!0,i.exports}t.m=e,t.c=n,t.d=function(e,n,r){t.o(e,n)||Object.defineProperty(e,n,{configurable:!1,enumerable:!0,get:r})},t.n=function(e){var n=e&&e.__esModule?function(){return e.default}:function(){return e};return t.d(n,"a",n),n},t.o=function(e,n){return ...
```
What happened here? I can't see what changed :(

kylehickinson replied:
This is the webpack'd JS (`npm run build` will webpack all the scripts in `Client/Frontend/UserContent/UserScripts/AllFrames/*` and `Client/Frontend/UserContent/UserScripts/MainFrame/*`

iccub commented on Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift:
```diff
@@ -93,7 +112,23 @@ extension BrowserViewController: RewardsDataSource {
+        let getHtmlToStringJSCall = "document.documentElement.outerHTML.toString()"
+        
+        DispatchQueue.main.async {
+            webView.evaluateJavaScript(getHtmlToStringJSCall, completionHandler: { html, _ in
```
we should log `error` here imo, unless this js call can throw a lot on unnecessary errors

Brandon-T commented on Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentStart/RewardsReporting.js:
```diff
@@ -0,0 +1,108 @@
+        });
+    }
+    
+    let originalOpen = null;
```
We should initialize some of these to what they are being set to later on?

IE: 
```javascript
let originalOpen = XMLHttpRequest.prototype.open;
let originalSend = XMLHttpRequest.prototype.send;
let originalFetch = window.fetch;
let originalSendBeacon = navigator.sendBeacon;
let originalImageSrc = Object.getOwnPropertyDescriptor(Image.prototype, "src");
```

kylehickinson replied:
Yeah we could do that too, basically just copy pasted from https://github.com/brave/brave-ios/blob/development/Client/Frontend/UserContent/UserScripts/MainFrame/AtDocumentStart/TrackingProtectionStats.js

iccub commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -147,13 +149,32 @@ class BrowserViewController: UIViewController {
         super.init(nibName: nil, bundle: nil)
         didInit()
+        
+        #if !NO_REWARDS
```
Can we extract it to `rewardsObserverSetup` method?

kylehickinson replied:
`rewardsObserverSetup`? If you mean the code above super.init: unfortunately not because  we reference self in the closures

iccub replied:
Below init, just moving these `fetchedPanelPublisher` and rest to a new separate method to not flood the initializer with code
Sorry for not being clear

kylehickinson replied:
Ah, sure yeah I can do that

Brandon-T commented on Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentStart/RewardsReporting.js:
```diff
@@ -0,0 +1,108 @@
+    
+    originalImageSrc = Object.getOwnPropertyDescriptor(Image.prototype, "src");
+    delete Image.prototype.src;
+    Object.defineProperty(Image.prototype, "src", {
```
Not sure if what I'm saying is necessary, but `HTMLImageElement` (aka `Image`) has 3 source properties:

```
src
lowsrc
currentSrc
```

If the page has a `media` element, it "might" choose a different source based on resolution and device. Does that in any way affect this code?

kylehickinson replied:
The simple answer to that is: I don't know :) 

It works for the set of publishers at the moment (YouTube specifically uses Image src's), and as far as I know none of the other publishers use it

iccub commented on Client/Frontend/Browser/Tab.swift:
```diff
@@ -151,7 +150,7 @@ class Tab: NSObject {
     init(configuration: WKWebViewConfiguration, type: TabType = .regular) {
         self.configuration = configuration
-        rewardsId = UInt32.random(in: UInt32.min...UInt32.max)
+        rewardsId = UInt32.random(in: 1...UInt32.max)
```
I feel that we should make a unit test and verify that a new tab will always have rewardsId >0 to catch regressions. I know it's a small thing, but we have to start somewhere

kylehickinson replied:
Im not entirely sure thats doable because its a random number generating between 1 and 4294967295, the test would pass 99.99999% of the, wouldn't it?

iccub replied:
Yeah I guess, and it's a let constant not a var so you are right

Brandon-T commented on Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentStart/RewardsReporting.js:
```diff
@@ -0,0 +1,108 @@
+    mutationObserver = new MutationObserver(function(mutations) {
+      mutations.forEach(function(mutation) {
+        mutation.addedNodes.forEach(function(node) {
+          if (node.tagName === "SCRIPT" && node.src) {
```
Alternatively:
`node.constructor.name == "script"`. But not necessary.

iccub commented on Client/Frontend/Browser/Tab.swift:
```diff
@@ -387,9 +386,8 @@ class Tab: NSObject {
-                rewards.reportLoadedPage(url: url, tabId: tabId, html: htmlString) { verified in
-                    self.tabDelegate?.tab(self, isVerifiedPublisher: verified)
-                }
+                let faviconURL = URL(string: self.displayFavicon?.url ?? "")
```
I would add a log here in case favicon can't be parsed to `URL`

kylehickinson replied:
What if there's just no favicon for this given tab?

iccub replied:
For default favicons it a website's url is set from what I see in `setIconURL`
`let defaults = defaultFavicon(url)`

kylehickinson replied:
Wouldn't that log an error for all Tab's using that default favicon then? Since theres no URL?

iccub replied:
I did some tests and for default favicons it just appends `/favicon.ico` to the url.

Example url without favicon `https://www.suchedniow.pl/`

```
// If we haven't got any options icons, then use the default at the root of the domain.
                if let url = NSURL(string: "/favicon.ico", relativeTo: url as URL), icons.isEmpty, let absoluteString = url.absoluteString {
                    let icon = Favicon(url: absoluteString)
                    icons = [icon]
                }
```

The only place where there's no faviconUrl is on new tab, reader mode website have faviconUrl as well.

I guess it's not a big deal to add or don't add those logs, we can always remove it after

kylehickinson replied:
Gotcha, thanks for checking into it :) I'll add the logs and re-request review after

iccub commented on Client/Frontend/Browser/TabManager.swift:
```diff
@@ -262,15 +256,12 @@ class TabManager: NSObject {
         if !PrivateBrowsingManager.shared.isPrivateBrowsing {
-            rewards.reportTabUpdated(Int(previousTab.rewardsId), url: previousTabUrl, isSelected: false,
-                                     isPrivate: previousTab.isPrivate) { _ in }
+            let faviconURL = URL(string: tab?.displayFavicon?.url ?? "")
```
same with adding a log for failed unwrapping

diracdeltas commented on Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift:
```diff
@@ -93,7 +112,26 @@ extension BrowserViewController: RewardsDataSource {
     
-    func displayString(for url: URL) -> String? {
-        return url.host
+    func pageHTML(for tabId: UInt64, completionHandler: @escaping (String?) -> Void) {
```
where does the result of this go? need to make sure it's not evaluated anywhere since page HTML is untrusted input

kylehickinson replied:
The HTML is pushed through to the native ledger and ads library. 
For ads its for classifying the page: https://github.com/brave/brave-core/blob/5a217978eee7639e7d8b9b755a4af11504698dda/vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc#L554
For ledger its used for twitch.tv publishers to scrape usernames I believe: https://github.com/brave/brave-core/blob/5a217978eee7639e7d8b9b755a4af11504698dda/vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc#L1031

diracdeltas replied:
got it

diracdeltas commented on Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentStart/RewardsReporting.js:
```diff
@@ -0,0 +1,100 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
does being in AllFrames mean this runs in iframes? i think we probably only want it for top-level pages?

kylehickinson replied:
Yes it does mean that, I'm not entirely sure actually if its required. Chromium's implementation has them overriding `OnBeforeRequest` (https://github.com/brave/brave-core/blob/5a217978eee7639e7d8b9b755a4af11504698dda/components/brave_rewards/browser/net/network_delegate_helper.cc#L80) and `ResourceDidLoad` (https://github.com/brave/brave-core/blob/5a217978eee7639e7d8b9b755a4af11504698dda/browser/brave_rewards/rewards_tab_helper.cc#L74)

So I assume it means that its executing for all frames

diracdeltas replied:
maybe it needs to be in all iframes to deal with media embeds?

diracdeltas replied:
actually can someone confirm that this does in fact need to run on all origins? would it be acceptable to only inject this content script on top-level twitch/youtube/vimeo/etc. pages?

kylehickinson replied:
@diracdeltas as far as I know it checks the URL in the native libs before processing anything that is reported, however it'd be difficult to only inject this content script based on the page because of the available WebKit API's

kylehickinson replied:
It would be nice to know though if this script can only be injected into main frame

diracdeltas commented on Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentStart/RewardsReporting.js:
```diff
@@ -0,0 +1,100 @@
+      }
+    });
+
+    mutationObserver = new MutationObserver(function(mutations) {
```
FWIW i think we got rid of mutation observers in browser-laptop a while ago due to performance issues

kylehickinson replied:
I'm not sure if there'll be a better way for us to do this because desktop reports all script loads: https://github.com/brave/brave-core/blob/5a217978eee7639e7d8b9b755a4af11504698dda/browser/brave_rewards/rewards_tab_helper.cc#L81

diracdeltas replied:
@NejcZdovc can you confirm why we need to check every script/XHR/etc. load in brave-core rewards to see if it is a media type?

i glanced through the code and it seems this is only necessary to check if the load is a twitch/youtube/vimeo request? 

i would rather we not override the prototypes and add mutation observers on every page just to see if they are twitch/youtube/vimeo loads

NejcZdovc replied:
we need to send every url into native ledger, which has business logic to check if twitch/youtube/etc. If we don't detect any media publisher we will ignore request

diracdeltas replied:
@NejcZdovc would it be acceptable to only send these requests if the top-level (or iframe-level) page is a twitch/youtube/etc. page? that would limit the potential impact of this change.

NejcZdovc replied:
problem is that then business logic is moved from native ledger (shared lib) to platform specific code. This means that if we add new publisher every platform need to add support for it instead of just pulling the latest shared lib

diracdeltas replied:
@NejcZdovc a somewhat separate question is - why does native ledger need to inspect all these subresource loads (XHRs, images) instead of just checking the top-level or iframe-level URL to see if it is twitch/youtube/etc?

NejcZdovc replied:
I already left comment in the code that I think we only need XHR and nothing else

kylehickinson replied:
I think you'll need at least Image as well for YouTube since that's how the load is classified for `watchtime`. Fetch and beacon API calls probably run through XHR in desktop for twitch and vimeo. As far as I've seen none of the publishers currently require script monitoring.

diracdeltas replied:
but AFAICT, we only need to report those XHRs on youtube.com and youtube embeds. so it seems preferable for me to only run this content script if the frame URL matches youtube/vimeo/twitch/etc. instead of on all pages.

i agree this is not ideal from a code-reuse perspective, but i think it's probably worth it to limit the potential webcompat (and maybe security) impacts of this overriding

kylehickinson commented:
@diracdeltas Updated to:
1. Not report anything from private tab's
2. Not report if rewards is disabled
3. Remove mutation observer, all currently supported media publishers still worked. We can re-add it at a later date if needed

Brandon-T commented:
> @diracdeltas Updated to:
> 
> 1. Not report anything from private tab's
> 2. Not report if rewards is disabled
> 3. Remove mutation observer, all currently supported media publishers still worked. We can re-add it at a later date if needed

We can always listen for when an element is constructed and if it’s a certain kind of element that you want to hook, you do your logic there in the future.

kylehickinson commented:
@diracdeltas Any updates on this PR? :)

diracdeltas commented:
sorry was waiting for a response from @NejcZdovc here: https://github.com/brave/brave-ios/pull/1387#discussion_r314487040

kylehickinson committed:
Fix #1386: Add rewards reporting to support media publishers and activity

- Panel now shows proper channels, names, etc. for supported publishers (YouTube, Vimeo, Twitch, GitHub, Twitter)
- Panel now shows proper activity for video related publishers (YouTube, Vimeo, Twitch)
- Now "hides" the tab (therefore computing activity) when the rewards panel is shown and "shows" it when its hidden
- Does the same when full-screen modals appear (tab is technically "hidden")
- Shows correct verified badge for publishers
- Supports notification count, however UI isn't added yet to show the red dot yet

kylehickinson committed:
Don't inject JS overrides for rewards unless on a media publisher origin

kylehickinson commented:
@diracdeltas Temporary solution implemented, will create a new ticket to replace hard-coded publisher origins with something better

diracdeltas commented on Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentStart/RewardsReporting.js:
```diff
@@ -0,0 +1,90 @@
+]
+
+if (mediaPublisherOrigins.includes(document.location.origin) && webkit.messageHandlers.rewardsReporting) {
+    install();
```
lgtm!

diracdeltas commented:
@kylehickinson thanks, this lgtm. i'll close out the sec review.

kylehickinson commented:
Security review concluded (https://github.com/brave/internal/issues/638)
